module/zfs/spa_errlog.c

```diff
@@ -128,6 +249,317 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+			error = dsl_dataset_hold_obj(dp, za.za_first_integer,
+			    FTAG, &clone);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+
+			boolean_t found = B_FALSE;
+			for (int i = 0; i < aff_snap_count; i++) {
+				if (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+				    == snap_obj_array[i]) {
+					found = B_TRUE;
+				}
+			}
+			dsl_dataset_rele(clone, FTAG);
+
+			if (!found)
+				continue;
+
+			error = check_filesystem(spa, za.za_first_integer, zep,
+			    count, addr, only_count);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+		}
+		zap_cursor_fini(&zc);
+	}
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (error);
+}
+
+static int
+find_top_affected_fs(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
+    uint64_t *top_affected_fs)
+{
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+
+	int error = dsl_dataset_hold_obj(dp, head_ds, FTAG, &ds);
+	if (error != 0)
+		return (error);
+
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	*top_affected_fs = head_ds;
+
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg) {
+		dsl_dataset_rele(ds, FTAG);
+		if (dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds) != 0)
+			break;
+		snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+		snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+		*top_affected_fs =
+		    dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj;
+	}
+	dsl_dataset_rele(ds, FTAG);
+	return (0);
+}
+
+static int
+process_error_block(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
+    uint64_t *count, void *addr, boolean_t only_count)
+{
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+	dsl_pool_config_enter(dp, FTAG);
+	uint64_t top_affected_fs;
+
+	int error = find_top_affected_fs(spa, head_ds, zep, &top_affected_fs);
+	if (error != 0)
+		goto out;
+
+	error = check_filesystem(spa, top_affected_fs, zep, count, addr,
+	    only_count);
+
+out:
+	dsl_pool_config_exit(dp, FTAG);
+	return (error);
+}
+
+static uint64_t
+get_errlog_size(spa_t *spa, uint64_t spa_err_obj)
+{
+	uint64_t total = 0;
+	if (spa_err_obj == 0)
+		return (total);
+
+	zap_cursor_t zc;
+	zap_attribute_t za;
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa_err_obj);
+	    zap_cursor_retrieve(&zc, &za) == 0; zap_cursor_advance(&zc)) {
+
+		zap_cursor_t head_ds_cursor;
+		zap_attribute_t head_ds_attr;
+		zbookmark_err_phys_t head_ds_block;
+
+		uint64_t head_ds;
+		name_to_object(za.za_name, &head_ds);
+
+		for (zap_cursor_init(&head_ds_cursor, spa->spa_meta_objset,
+		    za.za_first_integer); zap_cursor_retrieve(&head_ds_cursor,
+		    &head_ds_attr) == 0; zap_cursor_advance(&head_ds_cursor)) {
+
+			name_to_errphys(head_ds_attr.za_name, &head_ds_block);
+			if (process_error_block(spa, head_ds, &head_ds_block,
+			    &total, NULL, B_TRUE) != 0) {
+				zap_cursor_fini(&head_ds_cursor);
+				zap_cursor_fini(&zc);
+				return (total);
```

Comment:
`get_errlog_size()` is called by `spa_get_errlog_size()`, and both do not have the ability to return an error, meaning in case of errors we will end up with a partial total.

I think your suggestion to "drive on" by advancing the cursor makes more sense here as the total reported will be closer to the "real" value, had an error not occurred.

Reply:
OK. Are there any negative consequences of an incorrect total here? Where is this used? Is it just part of the reported output?

Reply:
I do not think there are any negative sequences of a partial total. All the upstream functions calling `spa_get_errlog_size()` return output to the user.

`spa_get_errlog_size()` is used:
1. in  `zpool_refresh_stats()` -> `zfs_ioc_pool_stats()` -> `spa_get_stats()`, ZPOOL_CONFIG_ERRCOUNT is populated. This is used for userspace reporting in `zpool_get_errlog()`, `zpool_do_status()` and `check_status()`.
2. in `zpool_get_errlog()` -> `zfs_ioc_error_log()` when `spa_get_errlog()` returns an error, it turns to `spa_get_errlog_size()` to determine the size of the error logs.
3. in `dsl_scan_done()` when logging in spa history

Reply:
OK, great